server/webapp/WEB-INF/rails.new/app/controllers/api_v1/admin/pluggable_scms_controller.rb

```diff
@@ -0,0 +1,86 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+##########################################################################
+
+module ApiV1
+  module Admin
+    class PluggableScmsController < ApiV1::BaseController
+      before_action :check_admin_user_and_401
+      before_action :check_for_stale_request, :check_for_scm_rename, only: [:update]
+
+
+      def index
+        scms = pluggable_scm_service.listAllScms
+        render DEFAULT_FORMAT => ApiV1::Scms::PluggableScmsRepresenter.new(scms).to_hash(url_builder: self)
+      end
+
+      def show
+        hash = get_scm_hash(params[:material_name])
+        render DEFAULT_FORMAT => hash if stale?(etag: JSON.generate(hash))
+      end
+
+      def create
+        result = HttpLocalizedOperationResult.new
+        @scm = ApiV1::Scms::PluggableScmRepresenter.new(SCM.new).from_hash(params)
+        @scm.ensureIdExists
+        pluggable_scm_service.createPluggableScmMaterial(current_user, @scm, result)
+
+        json = ApiV1::Scms::PluggableScmRepresenter.new(@scm).to_hash(url_builder: self)
+        result.isSuccessful ? (render DEFAULT_FORMAT => json) : (render_http_operation_result(result, {data: json}))
+      end
+
+      def update
+        result = HttpLocalizedOperationResult.new
+        @scm = ApiV1::Scms::PluggableScmRepresenter.new(SCM.new).from_hash(params)
+        pluggable_scm_service.updatePluggableScmMaterial(current_user, @scm, result)
+
+        json = ApiV1::Scms::PluggableScmRepresenter.new(@scm).to_hash(url_builder: self)
+        result.isSuccessful ? (render DEFAULT_FORMAT => json) : (render_http_operation_result(result, {data: json}))
+      end
+
+      def check_for_stale_request
+        if (request.env["HTTP_IF_MATCH"] != "\"#{get_etag_for_material(params[:material_name])}\"")
+          result = HttpLocalizedOperationResult.new
+          result.stale(LocalizedMessage::string("STALE_SCM_CONFIG", params[:material_name]))
+          render_http_operation_result(result)
+        end
+      end
+
+      def get_etag_for_material(material_name)
+        hash = get_scm_hash(material_name)
+        get_etag_for_scm_json(hash)
+      end
+
+      def get_etag_for_scm_json(hash)
+        Digest::MD5.hexdigest(JSON.generate(hash))
+      end
+
+      def check_for_scm_rename
+        unless params[:name].downcase == params[:material_name].downcase
+          result = HttpLocalizedOperationResult.new
+          result.notAcceptable(LocalizedMessage::string("GLOBAL_COPY_OF_SCM_CANNOT_BE_RENAMED"))
```

Comment:
Please understand the semantics of [http not-acceptable](https://httpstatuses.com/406)

> The target resource does not have a current representation that would be acceptable to the user agent, according to the proactive negotiation header fields received in the request1, and the server is unwilling to supply a default representation.

You really want to be using 422, and this is a simpler version of what you have here, with no need for localization.

```ruby
render_message('Renaming of SCM material is not supported by this API', :unprocessable_entity)
```